### ISSUE TITLE
Support password expiration

### ISSUE THREAD

**Issue by HaoK** [aspnet/AspNetCore#5716]

**Comment by harishakim** [comment]
Hi,Everybody 
please give more details about password expired date

**Comment by Janidbest** [comment]
PCI demands users to update the password and also its a best practice being enforced in most of the enterprise systems. Regardless of the business requirement, I guess this is most wanted feature for an identity framework. 

Any update please?

Does not want to duplicate code. ( which I already have two-factor implemented prior release). 

I did a quick implementation i.e.  `UserManager` - `CreateAsync` to add `PasswordExpiryEnabled` and `PasswordEnd` as per the configuration to AspNetUser. Then on login `signInManager.CheckPasswordSignInAsync` following a conditional check before `signInManager.SignInAsync`. 

This seems to work however its good to have this included in the framework.

**Comment by Janidbest** [comment]
There is another problem with the above approach mentioned to handle password expiration. 
 The problem is we can not have two-factor authentication. Because when we do a check on the VerifyCode page using the below , user always returns null. 

`var user = await signInManager.GetTwoFactorAuthenticationUserAsync();`
Thats because the user signin but not for twofactor signin. 
The method unfortunately private, and can't think of any work around.  
` private async Task<SignInResult> SignInOrTwoFactorAsync(TUser user, bool isPersistent, string loginProvider = null, bool bypassTwoFactor = false)`

So I guess we need the framework to have this functionality built in.

**Comment by Janidbest** [comment]
in the interim as a work around, I'm using the below (it seems to work for now)
````
  //handle two-factor authentication
                    if (user.TwoFactorEnabled )
                    {
                        await signInManager.PasswordSignInAsync(user, model.NewPassword, model.IsPersistent, true);
                        return await TwoFactorAuthenticationHandler(user, model.IsPersistent, model.ReturnUrl);
                    }
                    else {

                        await signInManager.SignInAsync(user, model.IsPersistent);
                        await appUserService.ManagePasswordExpiryAsync(user);                     

                        return RedirectToLocal(model.ReturnUrl);
                    }`
```

**Comment by valeriob** [comment]
Hi @blowdart ! Is there something in 3.0 preview already ?

**Comment by blowdart** [comment]
No, other things took precedence, and identity gets no extra features for 3.0, hence it going to backlog for now.

**Comment by fabich** [comment]
> PCI demands users to update the password and also its a best practice being enforced in most of the enterprise systems. Regardless of the business requirement, I guess this is most wanted feature for an identity framework.
> 
> Any update please?

@Janidbest there are more and more security papers and best-practice recommendations to no longer enforce password expiration.
e.g. https://www.sans.org/security-awareness-training/blog/time-password-expiration-die
or directly from Microsoft:
https://www.microsoft.com/en-us/research/publication/password-guidance

**Comment by stenionobres** [comment]
Hi! Is there any news about this feature?

**Comment by Xyncgas** [comment]
as the last issue currently standing in the repository, I press F for condolences

**Comment by Xyncgas** [comment]
IMO, this feature can be implemented by an intern taking minimum wage at a high school relationship tech company, under 30 min

**Comment by Xyncgas** [comment]
this can't be done on client side that doesn't keep a state anyways, just use a database or whatever write your own code to expire your password, heck just remove it on server and verify the email to reactivate the account with a new password
